server/catgenome/src/main/java/com/epam/catgenome/util/feature/reader/AbstractFeatureReader.java

```diff
@@ -0,0 +1,211 @@
+package com.epam.catgenome.util.feature.reader;
+
+/*
+ * Copyright (c) 2007-2010 by The Broad Institute, Inc. and the Massachusetts Institute of Technology.
+ * All Rights Reserved.
+ *
+ * This software is licensed under the terms of the GNU Lesser General Public License (LGPL), Version 2.1 which
+ * is available at http://www.opensource.org/licenses/lgpl-2.1.php.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS." THE BROAD AND MIT MAKE NO REPRESENTATIONS OR WARRANTIES OF
+ * ANY KIND CONCERNING THE SOFTWARE, EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION, WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NONINFRINGEMENT, OR THE ABSENCE OF LATENT
+ * OR OTHER DEFECTS, WHETHER OR NOT DISCOVERABLE.  IN NO EVENT SHALL THE BROAD OR MIT, OR THEIR
+ * RESPECTIVE TRUSTEES, DIRECTORS, OFFICERS, EMPLOYEES, AND AFFILIATES BE LIABLE FOR ANY DAMAGES OF
+ * ANY KIND, INCLUDING, WITHOUT LIMITATION, INCIDENTAL OR CONSEQUENTIAL DAMAGES, ECONOMIC
+ * DAMAGES OR INJURY TO PROPERTY AND LOST PROFITS, REGARDLESS OF WHETHER THE BROAD OR MIT SHALL
+ * BE ADVISED, SHALL HAVE OTHER REASON TO KNOW, OR IN FACT SHALL KNOW OF THE POSSIBILITY OF THE
+ * FOREGOING.
+ */
+
+import com.epam.catgenome.util.Utils;
+import htsjdk.tribble.*;
+import htsjdk.tribble.index.Index;
+import htsjdk.tribble.util.ParsingUtils;
+import htsjdk.tribble.util.TabixUtils;
+
+import java.io.File;
+import java.io.IOException;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * jrobinso
+ * <p/>
+ * the feature reader class, which uses indices and codecs to read in Tribble file formats.
+ */
+public abstract class AbstractFeatureReader<T extends Feature, S> implements FeatureReader<T> {
```

Comment:
The same here, an inheritance from HTSJDK's class might be a better solution, if it is possible

Reply:
We need the field `path` for `readIndex()`, but it is package-private, so it can't be used in sub-sub-classes (like subclasses of `TabixReader`).

Reply:
Okay, got it